pkg/cloudprovider/providers/aws/volumecache.go

```diff
@@ -0,0 +1,265 @@
+			} else {
+				glog.V(2).Infof("triggering volume refresh because state expired (volume %s, %s)", i.ID, cachePolicy.Name)
+			}
+		}
+	}
+
+	updated, err := i.cache.refreshAll(snapshot)
+	if err != nil {
+		return nil, err
+	}
+
+	instance := updated.findCachedVolume(i.ID)
+	if instance == nil {
+		return nil, cloudprovider.InstanceNotFound
+	}
+
+	state := i.findStateIfCached(nil)
+	if state != nil {
+		return state, nil
+	}
+
+	return nil, cloudprovider.InstanceNotFound
+}
+
+// findStateIfCached returns the last state, if it is valid under the cache policy.
+// If there is no state, or the state has expired, it returns nil.
+func (i *cachedVolume) findStateIfCached(cachePolicy *CachePolicy) *ec2.Volume {
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+
+	if i.state != nil {
+		if cachePolicy == nil || cachePolicy.IsValid(i.stateTimestamp) {
+			return i.state
+		} else {
+			glog.V(2).Infof("triggering refresh because cache policy %s expired", cachePolicy.Name)
+		}
+	}
+
+	return nil
+}
+
+// newInstanceCache is a constructor for an volumeCache
+func newVolumeCache(ec2 EC2) *volumeCache {
+	return &volumeCache{
+		ec2: ec2,
+	}
+}
+
+// latestSnapshot returns the last view of all volumes (or nil if no snapshot)
+func (c *volumeCache) latestSnapshot() *volumeCacheSnapshot {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	return c.latest
+}
+
+// refreshAll builds an updated snapshot by calling DescribeVolume.
+// It accepts the previous snapshot, and will not refresh if a concurrent request has triggered a concurrent refresh.
+func (c *volumeCache) refreshAll(previous *volumeCacheSnapshot) (*volumeCacheSnapshot, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.latest != nil && c.latest != previous {
+		// A concurrent refresh happened, return that
+		return c.latest, nil
+	}
+
+	request := &ec2.DescribeVolumesInput{}
+
+	volumes, err := c.ec2.DescribeVolumes(request)
+	if err != nil {
+		glog.V(2).Infof("error querying EC2 volumes: %v", err)
+		return nil, err
+	}
+
+	byVolumeId := make(map[awsVolumeID]*cachedVolume)
+
+	now := nanoTime()
+
+	latest := c.latest
+	if latest == nil {
+		// Easier than nil checks everywhere
+		latest = &volumeCacheSnapshot{}
+	}
+
+	for _, volumeDetails := range volumes {
+		volumeID := awsVolumeID(aws.StringValue(volumeDetails.VolumeId))
+
+		if volumeID == "" {
+			glog.Warningf("ignoring volume without id: %v", volumeDetails)
+			continue
+		}
+
+		if byVolumeId[volumeID] != nil {
+			return nil, fmt.Errorf("multiple volumes found for volume: %s", volumeID)
+		}
+
+		ci := latest.byVolumeId[volumeID]
+
+		if ci == nil {
+			az := ""
+			if volumeDetails.AvailabilityZone != nil {
+				az = aws.StringValue(volumeDetails.AvailabilityZone)
+			}
+
+			ci = &cachedVolume{
+				ID:               volumeID,
+				AvailabilityZone: az,
+
+				cache: c,
+			}
+		}
+
+		byVolumeId[volumeID] = ci
+		{
+			ci.mutex.Lock()
+			ci.state = volumeDetails
+			ci.stateTimestamp = now
+			ci.mutex.Unlock()
+		}
```

Comment:
Why these locks are here btw? Is this to prevent from some other piece of code accidently setting `state` of `CachedInstance`? 

But if another goroutine already has access to a pointer to `ci` then this lock in itself won't prevent the mutation of cached instance.